src/JSON/Serializer.dfy

```diff
@@ -0,0 +1,177 @@
+/// =============================================
+///  Serialization from JSON.AST to JSON.Grammar
+/// =============================================
+///
+/// For the spec, see ``JSON.Spec.dfy``.
+
+include "../Collections/Sequences/Seq.dfy"
+include "../BoundedInts.dfy"
+include "../Math.dfy"
+
+include "Utils/Views.dfy"
+include "Utils/Vectors.dfy"
+include "Utils/Unicode.dfy"
+include "Errors.dfy"
+include "AST.dfy"
+include "Grammar.dfy"
+include "Spec.dfy"
+
+module {:options "-functionSyntax:4"} JSON.Serializer {
+  import Seq
+  import Math
+  import opened Wrappers
+  import opened BoundedInts
+  import opened Utils.Str
+  import Utils.Unicode
+
+  import AST
+  import Spec
+  import opened Errors
+  import opened Utils.Vectors
+  import opened Grammar
+  import opened Utils.Views.Core
+
+  type Result<+T> = SerializationResult<T>
+
+  type bytes = seq<uint8>
+  type bytes32 = bs: bytes | |bs| < TWO_TO_THE_32
+  type string32 = s: string | |s| < TWO_TO_THE_32
+
+  function Bool(b: bool): jbool {
+    View.OfBytes(if b then TRUE else FALSE)
+  }
+
+  function Transcode16To8Escaped(str: string, start: uint32 := 0): bytes {
+    Unicode.Transcode16To8(Spec.Escape(str))
+  } // FIXME speed up using a `by method`
+  // by method {
+  //   var len := |str| as uint32;
+  //   if len == 0 {
+  //     return [];
+  //   }
+  //   var st := new Vectors.Vector(0, len);
+  //   var c0: uint16 := 0;
+  //   var c1: uint16 := str[0] as uint16;
+  //   var idx: uint32 := 0;
+  //   while idx < len {
+  //     var c0 := c1;
+  //     var c1 := str[idx + 1] as uint16;
+  //     if c0 < 0xD800 || c0 > 0xDBFF {
+  //       Utf8Encode(st, Unicode.Utf16Decode1(c0));
+  //       idx := idx +1;
+  //     } else {
+  //       Utf8Encode(st, Unicode.Utf16Decode2(c0, c1));
+  //       idx := idx + 2;
+  //     }
+  //   }
+  // }
+
+  function CheckLength<T>(s: seq<T>, err: SerializationError): Outcome<SerializationError> {
+    Need(|s| < TWO_TO_THE_32, err)
+  }
+
+  function String(str: string): Result<jstring> {
+    var bs := Transcode16To8Escaped(str);
+    :- CheckLength(bs, StringTooLong(str));
+    Success(Grammar.JString(Grammar.DOUBLEQUOTE, View.OfBytes(bs), Grammar.DOUBLEQUOTE))
+  }
+
+  function Sign(n: int): jminus {
+    View.OfBytes(if n < 0 then ['-' as byte] else [])
+  }
+
+  module ByteStrConversion refines Str.ParametricConversion {
+    import opened BoundedInts
+    type Char = uint8
+  }
+
+  const DIGITS := [
+    '0' as uint8, '1' as uint8, '2' as uint8, '3' as uint8,
+    '4' as uint8, '5' as uint8, '6' as uint8, '7' as uint8,
+    '8' as uint8, '9' as uint8
+  ]
+
+  const MINUS := '-' as uint8
+
+  function Int'(n: int) : (str: bytes)
+    ensures forall c | c in str :: c in DIGITS || c == MINUS
```

Comment:
I don't want to do this change yet, because the rest of the code is compatible with older Dafny.  Once we switch to this syntax as default we can make this change.